### PR TITLE
Data models and data ingestion jobs

### PR DESCRIPTION
**Pull request by Anakin100100**
Now we are able to take the data from RSPO api an ingest it into the database as appropriate models. The specific filtering strategy hasn't been implemented yet.

### CODE REVIEW

**Comment by micorix** [comment]
Hi! Reminder to use imperative mood when wording commit messages and PR names :slightly_smiling_face: 

Just 2 things to consider:
1. Maybe facility instead of institution? but it's just a suggestion, maybe it doesn't matter.
2. Can you include some basic docs on how to run things? We can ofc add them later but even short CLI instruction would be cool even now i think.

File: app/jobs/create_institution_records_job.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+      raw_institutions = JSON.parse(response.body)
+
+      # the records are sent in pages of 100
+      break if raw_institutions == []
+
+      i += 1
+
+      batch_insert_institutions = []
+      raw_institutions.each do |raw_institution|
+        institution = {}
```
Comment (wojtodzio) [inline app/jobs/create_institution_records_job.rb:27]: You can slightly shorten it in an idiomatic way using each_with_object
```suggestion
      raw_institutions.each_with_object({}) do |raw_institution, institution|
```

File: app/jobs/create_institution_records_job.rb [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+require 'net/http'
+
+class CreateInstitutionRecordsJob < ApplicationJob
+  queue_as :default
+
+  def perform(institution_type_id)
+    i = 1
+    loop do
+      rspo_api_uri = URI("http://194.54.26.132/api/placowki/?page=#{i}&typ_podmiotu_id=#{institution_type_id}")
```
Comment (wojtodzio) [inline app/jobs/create_institution_records_job.rb:11]: Have you ever heard of clients & wrappers? They would fit perfectly here: https://medium.com/selleo/essential-rubyonrails-patterns-clients-and-wrappers-c19320bcda0
That will allow you to reuse api-related logic throughout the application in an elegant way, and it will make it way easier to test this code

Also, for testing, you can use https://github.com/bblimke/webmock or https://github.com/vcr/vcr

File: app/jobs/create_institution_records_job.rb [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        institution['institution_type_id'] = raw_institution['typ']['id']
+        institution['name'] = raw_institution['nazwa']
+        institution['rspo_institution_id'] = raw_institution['numerRspo']
+        institution['public'] = raw_institution['statusPublicznoPrawny']['nazwa'] == 'publiczna'
+        institution['latitude'] = raw_institution['geolokalizacja']['latitude'].to_f
+        institution['longitude'] = raw_institution['geolokalizacja']['longitude'].to_f
+        institution['website'] = raw_institution['stronaInternetowa']
+        institution['email'] = raw_institution['email']
+        # rubocop:disable Layout/LineLength
+        institution['address'] = "#{raw_institution['adresDoKorespondecjiMiejscowosc']} #{raw_institution['adresDoKorespondecjiUlica']} #{raw_institution['adresDoKorespondecjiNumerBudynku']}/#{raw_institution['adresDoKorespondecjiNumerLokalu']} #{raw_institution['adresDoKorespondecjiKodPocztowy']}"
```
Comment (wojtodzio) [inline app/jobs/create_institution_records_job.rb:37]: If you're expecting a hash always to include a specific value, it's a good idea to use [`Hash#fetch`](https://apidock.com/ruby/Hash/fetch)
Now, if you make a mistake in the key's name, it will insert an empty value in the DB, which can be easy to miss. `fetch` will throw an exception, which is impossible to ignore. It will also immediately notify you if the API's response changes.

File: app/jobs/create_institution_records_job.rb [hunk 4]
```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+
+require 'net/http'
+
+class CreateInstitutionRecordsJob < ApplicationJob
+  queue_as :default
+
+  def perform(institution_type_id)
+    i = 1
```
Comment (wojtodzio) [inline app/jobs/create_institution_records_job.rb:9]: `i` isn't very descriptive. Consider using `page`

File: app/jobs/create_institution_records_job.rb [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+      # the records are sent in pages of 100
+      break if raw_institutions == []
+
+      i += 1
+
+      batch_insert_institutions = []
+      raw_institutions.each do |raw_institution|
+        institution = {}
+        institution['institution_type_id'] = raw_institution['typ']['id']
```
Comment (wojtodzio) [inline app/jobs/create_institution_records_job.rb:28]: Not a biggie, but you should be able to use symbol keys with the active record, and symbols are idiomatic to ruby

File: app/jobs/create_institution_records_job.rb [hunk 6]
```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+
+require 'net/http'
+
+class CreateInstitutionRecordsJob < ApplicationJob
```
Comment (wojtodzio) [inline app/jobs/create_institution_records_job.rb:5]: Jobs aren't perfect for storing business logic - they're rather hard to test, their logic is hard to reuse, and they are hard to refactor. It's rather customary for jobs to have a very minimal `perform` method and store the logic somewhere else - here, it'd probably be a service object since it's suitable for encapsulating side effects (example: https://github.com/Selleo/pattern/#service).
```ruby
def perform(institution_type_id)
  CreateInstitutionRecords.call(institution_type_id)
end
```
That way, it will be way easier for you to test `CreateInstitutionRecords` in isolation and reuse it in other places. You'll also be able to use instance variables and so do a proper object-oriented modeling

File: app/jobs/enque_parallel_institution_create_job.rb [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+class EnqueParallelInstitutionCreateJob < ApplicationJob
+  queue_as :default
+
+  def perform(*_args)
+    if Institution.count != 0
+      Rails.logger.debug 'To execute this job there can be no Institution records in the database'
+      return
+    end
+
+    relevant_institution_types_ids = [93, 94, 45, 96, 14, 15, 17, 15, 27, 26, 20, 16]
```
Comment (wojtodzio) [inline app/jobs/enque_parallel_institution_create_job.rb:12]: These seem like some magic numbers to me. Would be great to comment them somehow. I suggest assigning them to named consts - that way it'll be very easy to understand what they mean and to reuse them later if needed. Also, comments or documentation tends to get outdated very easily, while named variables do not, and so we end with self-documenting code. Something like
```ruby
PRIMARY_SCHOOL = 1
ANOTHER_ASYLUM = 2
...
```

File: app/models/institution_type.rb [hunk 8]
```diff
@@ -0,0 +1,6 @@
+# frozen_string_literal: true
+
+class InstitutionType < ApplicationRecord
+  # create a has many relationship with the institution model
```
Comment (wojtodzio) [inline app/models/institution_type.rb:4]: IMO this comment does not bring any value - `has_many :institutions` is self-documenting enough, no need to add any additional comments

File: config/database.yml [hunk 9]
```diff
@@ -2,7 +2,7 @@ default: &default
   adapter:  postgresql
   host:     localhost
   encoding: unicode
-  pool:     10
+  pool:     20
```
Comment (wojtodzio) [inline config/database.yml:5]: Why would you increase the pool? Usually, unless you use an external connection pooler, increasing the pool's size leads to poorer performance. As far as I see, Good Job does not recommend it either https://github.com/bensheldon/good_job#database-connections

File: config/environments/development.rb [hunk 10]
```diff
@@ -13,6 +13,8 @@
   # Do not eager load code on boot.
   config.eager_load = false
 
+  config.hosts.clear
```
Comment (wojtodzio) [inline config/environments/development.rb:16]: Why did you need to add this? May be worth adding a comment for posterity

File: config/puma.rb [hunk 11]
```diff
@@ -6,7 +6,7 @@
 # the maximum value specified for Puma. Default is set to 5 threads for minimum
 # and maximum; this matches the default thread size of Active Record.
 #
-max_threads_count = ENV.fetch('RAILS_MAX_THREADS', 5)
+max_threads_count = ENV.fetch('RAILS_MAX_THREADS', 20)
```
Comment (wojtodzio) [inline config/puma.rb:9]: 20 threads is a lot, do we really need that much?

File: config/routes.rb [hunk 12]
```diff
@@ -7,4 +7,5 @@
   # root "articles#index"
   get 'schools/get_all_schools', to: 'schools#all_schools'
   get 'schools/get_one_school', to: 'schools#one_school'
-end
+  mount GoodJob::Engine => 'good_job'
+end
```
Comment (wojtodzio) [inline config/routes.rb:11]: Missing newline

File: db/migrate/20220128071908_create_institution_types.rb [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+      t.integer :rspo_institution_id
+      t.string :name
+      t.boolean :public 
+      t.decimal :latitude, :precision=>10, :scale=>6
+      t.decimal :longitude, :precision=>10, :scale=>6
+      t.string :website 
+      t.string :email
+      t.string :address  
+      t.timestamps
+    end
```
Comment (wojtodzio) [inline db/migrate/20220128071908_create_institution_types.rb:20]: Can any of those columns be nil? If not, please add not-null constraint

File: db/migrate/20220128071908_create_institution_types.rb [hunk 14]
```diff
@@ -0,0 +1,22 @@
+class CreateInstitutionTypes < ActiveRecord::Migration[7.0]
+  def change
+    create_table :institution_types do |t|
+      t.string :name 
+      t.integer :rspo_institution_type_id
+      t.timestamps
+    end
+
+    create_table :institutions do |t|
+      t.belongs_to :institution_type
```
Comment (wojtodzio) [inline db/migrate/20220128071908_create_institution_types.rb:10]: Please add a foreign key constraint

File: db/migrate/20220128071908_create_institution_types.rb [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+      t.integer :rspo_institution_id
+      t.string :name
+      t.boolean :public 
+      t.decimal :latitude, :precision=>10, :scale=>6
+      t.decimal :longitude, :precision=>10, :scale=>6
```
Comment (wojtodzio) [inline db/migrate/20220128071908_create_institution_types.rb:15]: Please use the new hash format (`precision: 10` vs `:precision => 10`)

File: spec/jobs/create_institution_records_job_spec.rb [hunk 16]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe CreateInstitutionRecordsJob, type: :job do
+  pending "add some examples to (or delete) #{__FILE__}"
```
Comment (wojtodzio) [inline spec/jobs/create_institution_records_job_spec.rb:6]: 😢

File: spec/jobs/create_institution_types_job_spec.rb [hunk 17]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe CreateInstitutionTypesJob, type: :job do
+  describe 'When enquiuing a CreateInstitutionTypesJob' do
+    it 'adds all institution types to the database if there are no InstitutionType records' do
+      InstitutionType.delete_all
```
Comment (wojtodzio) [inline spec/jobs/create_institution_types_job_spec.rb:8]: DB should be empty at the beginning of every test

**Review by wojtodzio — CHANGES_REQUESTED** [review]
Pretty nice job! I'm leaving a few comments, but I'm being really nitpicky here

File: Dockerfile [hunk 18]
```diff
@@ -6,4 +6,4 @@ RUN apk add --virtual build-dependencies build-base ruby-dev libpq postgresql-de
 COPY Gemfile /code
 RUN bundle install 
 COPY . /code
-CMD rails s -b 0.0.0.0 -p ${RAILS_PORT:-80} -e production
+CMD rails s -b 0.0.0.0 -p ${RAILS_PORT:-80} -e production
```
[the remark this answers is not in the document]
Comment (Anakin100100) [reply Dockerfile:9]: is this a problem?

[the remark this answers is not in the document]
Comment (wojtodzio) [reply Dockerfile:9]: Missing newline can break some tools: https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline
You should configure your editor to automatically include the newline

File: app/jobs/create_institution_records_job.rb (hunk 2 above)
Comment (Anakin100100) [reply app/jobs/create_institution_records_job.rb:11]: I heard of them before but I never had the opportunity do design one. I don't think it is particularly necessary here because we are making only two distinct calls to this api and we are unlikely to need more in the near future.

File: config/environments/development.rb (hunk 10 above)
Comment (Anakin100100) [reply config/environments/development.rb:16]: This is a solution that I found to an error resulting from my computer not being on the allowed hosts list

File: app/jobs/create_institution_records_job.rb (hunk 2 above)
Comment (wojtodzio) [reply app/jobs/create_institution_records_job.rb:11]: As long as you can write good tests for that without introducing a client, I'm okay with that - although IMO clients are pretty easy to write, and they make it way easier to test the integration

File: app/jobs/create_institution_records_job.rb (hunk 6 above)
Comment (Anakin100100) [reply app/jobs/create_institution_records_job.rb:6]: I will design the data ingestion job from gdynia as a service object but lets leave these two as they are for now

File: config/routes.rb (hunk 12 above)
Comment (wojtodzio) [reply config/routes.rb:13]: still missing

File: db/migrate/20220128071908_create_institution_types.rb [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+    end
+
+    create_table :institutions do |t|
+      t.belongs_to :institution_type
+      t.integer :rspo_institution_id
+      t.integer :rspo_institution_type_id
+      t.string :name, null: false
+      t.boolean :public, null: false
+      t.decimal :latitude, precision: 10, :scale=>6, null: false
+      t.decimal :longitude, precision: 10, :scale=>6, null: false
```
Comment (wojtodzio) [inline db/migrate/20220128071908_create_institution_types.rb:16]:
```suggestion
      t.decimal :latitude, precision: 10, scale: 6, null: false
      t.decimal :longitude, precision: 10, scale: 6, null: false
```

File: db/migrate/20220128071908_create_institution_types.rb (hunk 14 above)
Comment (wojtodzio) [reply db/migrate/20220128071908_create_institution_types.rb:10]: Still missing
```suggestion
      t.belongs_to :institution_type, foreign_key: true, null: false
```

File: db/migrate/20220128071908_create_institution_types.rb [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+    create_table :institution_types do |t|
+      t.string :name 
+      t.integer :rspo_institution_type_id
+      t.timestamps
+    end
+
+    create_table :institutions do |t|
+      t.belongs_to :institution_type
+      t.integer :rspo_institution_id
+      t.integer :rspo_institution_type_id
```
Comment (wojtodzio) [inline db/migrate/20220128071908_create_institution_types.rb:12]: Can those be null?

File: db/migrate/20220130154129_rspo_foreign_key.rb [hunk 21]
```diff
@@ -0,0 +1,6 @@
+class RspoForeignKey < ActiveRecord::Migration[7.0]
+  def change
+    add_index :institution_types, [:rspo_institution_type_id], :unique => true
```
Comment (wojtodzio) [inline db/migrate/20220130154129_rspo_foreign_key.rb:3]: As long as your branch was not deployed yet, you can just rollback last migration locally and edit it

File: db/migrate/20220128071908_create_institution_types.rb [hunk 22]
```diff
@@ -0,0 +1,24 @@
+class CreateInstitutionTypes < ActiveRecord::Migration[7.0]
+  def change
+    create_table :institution_types do |t|
+      t.string :name 
+      t.integer :rspo_institution_type_id
```
Comment (wojtodzio) [inline db/migrate/20220128071908_create_institution_types.rb:5]: Can it be null?

File: spec/factories/institution_factory.rb [hunk 23]
```diff
@@ -0,0 +1,20 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
+  factory :institution do
+    name { Faker::FunnyName.name }
```
Comment (wojtodzio) [inline spec/factories/institution_factory.rb:5]: Up to you, but IMO, faker in tests brings very little value, and it makes it way harder to debug failing tests, as every time a test fails, you'll get a different error message. I prefer to either hardcode factory's values or use a sequence

File: spec/models/institution_spec.rb [hunk 24]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe Institution, type: :model do
+  describe 'methods' do
+    it "has a full address method from it's components" do
+      type = create(:institution_type)
+      i = create(:institution, institution_type: type)
```
Comment (wojtodzio) [inline spec/models/institution_spec.rb:9]: `i` is a terrible variable name, readability is one of the most important aspects of a well written and maintainable code
```suggestion
      institution = create(:institution, institution_type: type)
```

File: spec/models/institution_spec.rb [hunk 25]
```diff
@@ -0,0 +1,13 @@
...
+      expect(i.address).to eq("#{i.city} #{i.street} #{i.building_no}/#{i.apartment_no} #{i.zip_code}")
```
Comment (wojtodzio) [inline spec/models/institution_spec.rb:10]: Here you're basically reusing and testing the implementation.
Tests should be dumb (vs. smart) and, above all, readable. It'd be better to explicitly pass values to the factory and then use an actual output string, something like:
```ruby
institution = create(:institution, city: 'Bielsko', street: 'Lompy', ...)
expect(institution.address).to eq("Bielsko Lompy ...")
```
That way, you're describing and testing the specification instead of the implementation, and it's immediately clear how does the address format look like

File: spec/requests/get_one_institution_spec.rb [hunk 26]
```diff
@@ -0,0 +1,29 @@
...
+
+require 'rails_helper'
+
+RSpec.describe 'Institutions', type: :request do
+  describe 'GET /institutions/get_one_institution' do
+    before do
+      type = create(:institution_type)
+      (1..60).each do |_n|
+        create(:institution, institution_type: type)
+      end
```
Comment (wojtodzio) [inline spec/requests/get_one_institution_spec.rb:11]: You can use create_list instead of an explicit loop
```suggestion
      create_list(:institution, 60, institution_type: type)
```
Do you need 60 records in each example here, though? This approach will eventually make tests pretty slow and annoying

**Review by wojtodzio — CHANGES_REQUESTED** [review]
Almost there, probably the last round of review :)

File: spec/factories/institution_factory.rb (hunk 23 above)
Comment (Anakin100100) [reply spec/factories/institution_factory.rb:5]: ok but in some tests I need to generate lets say 20 institutions. How do i then use it? Hardcoding 20 names seems excessive. Can it be left to faker as default but overridden in some tests where it is useful? I think that leaving this as a default is fine.

File: spec/models/institution_spec.rb (hunk 24 above)
Comment (Anakin100100) [reply spec/models/institution_spec.rb:9]: I know, this was the previous variable name but after rubocop screamed that the line is too long.

File: config/routes.rb [hunk 27]
```diff
@@ -5,6 +5,9 @@
   # Define your application routes per the DSL in https://guides.rubyonrails.org/routing.html
   # Defines the root path route ("/")
   # root "articles#index"
-  get 'schools/get_all_schools', to: 'schools#all_schools'
-  get 'schools/get_one_school', to: 'schools#one_school'
-end
+  get 'institutions/get_all_institutions', to: 'institutions#all_institutions'
+  get 'institutions/get_one_institution', to: 'institutions#one_institution'
```
[the remark this answers is not in the document]
Comment (wojtodzio) [reply config/routes.rb:9]: Still not fixed

File: db/migrate/20220128071908_create_institution_types.rb (hunk 20 above)
Comment (wojtodzio) [reply db/migrate/20220128071908_create_institution_types.rb:12]: Still not fixed

File: db/migrate/20220131083110_create_gdynia_extra_data.rb [hunk 28]
```diff
@@ -0,0 +1,8 @@
+class CreateGdyniaExtraData < ActiveRecord::Migration[7.0]
+  def change
+    create_table :gdynia_extra_data do |t|
```
Comment (wojtodzio) [inline db/migrate/20220131083110_create_gdynia_extra_data.rb:3]: I'm not sure if having city-specific tables is a good idea, but I guess this is just a stub for future development?

File: db/schema.rb [hunk 29]
```diff
@@ -45,10 +50,32 @@
     t.index ["scheduled_at"], name: "index_good_jobs_on_scheduled_at", where: "(finished_at IS NULL)"
   end
 
-  create_table "schools", force: :cascade do |t|
+  create_table "institution_types", force: :cascade do |t|
     t.string "name"
+    t.integer "rspo_institution_type_id"
```
Comment (wojtodzio) [inline db/schema.rb:55]: Those probably cannot be null

File: db/schema.rb [hunk 30]
```diff
@@ -45,10 +50,32 @@
...
+    t.datetime "created_at", precision: 6, null: false
+    t.datetime "updated_at", precision: 6, null: false
+    t.index ["rspo_institution_type_id"], name: "index_institution_types_on_rspo_institution_type_id", unique: true
+  end
+
+  create_table "institutions", force: :cascade do |t|
+    t.bigint "institution_type_id"
+    t.integer "rspo_institution_id"
+    t.integer "rspo_institution_type_id"
```
Comment (wojtodzio) [inline db/schema.rb:64]: Those probably cannot be null

**Review by wojtodzio — CHANGES_REQUESTED** [review]
Just last few missed comments, almost there!

File: db/migrate/20220131083110_create_gdynia_extra_data.rb (hunk 28 above)
Comment (Anakin100100) [reply db/migrate/20220131083110_create_gdynia_extra_data.rb:3]: Pretty much so. For now this data comes only from one city and looking how hard to calculate are most of these factors we are unlikely to get any more data in this format. In case we do it can be reinterpreted as a data format that originates from Gdynia

File: spec/factories/institution_factory.rb (hunk 23 above)
Comment (wojtodzio) [reply spec/factories/institution_factory.rb:5]: You can use `sequence` if you need to have unique value in the database
```suggestion
  factory :institution do
    sequence(:name) { |n| "LO #{n}" }
```

Comment (wojtodzio) [reply spec/factories/institution_factory.rb:5]: BTW, if you want me to add some additional clarification to an old comment, please do not resolve it. I'm usually not checking resolved comments, and I'll probably miss those followups

File: db/schema.rb (hunk 29 above)
Comment (Anakin100100) [reply db/schema.rb:55]: yes and there is a foreign key placed on this field that guarantees null safety

File: db/schema.rb (hunk 30 above)
Comment (Anakin100100) [reply db/schema.rb:64]: they are set as not null in the current migration that creates them but these changes weren't applied to the schema yet.

**Review by wojtodzio — APPROVED** [review]
Really, really nice job! 🚀

File: db/schema.rb (hunk 29 above)
Comment (wojtodzio) [reply db/schema.rb]: I see there're not null constraints in schema now, but just FYI, foreign key constraints ensure reference consistency, but they're separate from not null constraints. If you wanted to have an optional association, you'd add a foreign key constraint, but you wouldn't add a not-null constraint.
